src/array.cr

```diff
@@ -1545,11 +1544,47 @@ class Array(T)
     self
   end
 
+  # Returns `self` with all the elements shifted `n` times.
+  #
+  # ```
+  # a1 = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+  # a2 = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+  # a3 = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+  #
+  # a1.rotate!
+  # a2.rotate!(1)
+  # a3.rotate!(3)
+  #
+  # a1 # => [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]
+  # a2 # => [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]
+  # a3 # => [3, 4, 5, 6, 7, 8, 9, 0, 1, 2]
+  # ```
   def rotate!(n = 1)
     return self if size == 0
+    # This is broken into cases for perfromance (#8515)
```

Comment:
I wouldn't mention a GitHub issue in code (in bug specs it's fine to understand why a spec for a specific case exists).

Reply:
Why not? What's the downside? Such a reference obviously provides context information. Worst case is when the repo switches away from GitHub, the reference might not lead anywhere. But apart from that...?

Reply:
I guess this is fine.

Reply:
We're actually using quite a few references to GitHub issues in code comments. I think that's fine.

Reply:
@asterite I misread your comment thinking that you're for it ;) we do it in specs and in this case would be pretty informative into _why_ exactly these optimizations were performed.

Reply:
> why exactly these optimizations were performed

I think optimizations are performed because they are always good? I don't think we need to say "Here's an optimization because it's better to have fast code instead of slow code" or "Here we are doing an optimization because someone realized this could be optimized". Or what could another reason for optimizing code be?

Reply:
Sure, optimizations are good in general. But there always are different possibilities and you wouldn't want to just randomly shoot optimizations around. Some work better than others, but there are always trade-offs and complex decisions why to choose a specific optimization.